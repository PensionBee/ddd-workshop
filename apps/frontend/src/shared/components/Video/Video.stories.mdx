import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import { Video } from "./Video";

<Meta
  title="Shared/Video/Docs"
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Video

## Props

<ArgsTable of={Video} />

## Usage & Customisation

### Default

<Canvas>
  <Story id="shared-video--default" />
</Canvas>

### Modal

<Canvas>
  <Story id="shared-video--modal" />
</Canvas>

### In column

<Canvas>
  <Story id="shared-video--in-column" />
</Canvas>

### Muted

<Canvas>
  <Story id="shared-video--muted" />
</Canvas>

### No thumbnail

<Canvas>
  <Story id="shared-video--no-thumbnail" />
</Canvas>
