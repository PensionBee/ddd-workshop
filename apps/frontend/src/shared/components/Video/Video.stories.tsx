import { Box } from "../Box/Box";
import type { ComponentMeta } from "@storybook/react";
import { Video } from "./Video";
import { GridContainer } from "../Grid";
import { GridItem } from "../Grid/GridItem/GridItem";
import { GridRow } from "../Grid/GridRow/GridRow";

export default {
  title: "Shared/Video",
  component: Video,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof Video>;

export const Default = () => {
  return (
    <GridContainer>
      <Video
        src="https://res.cloudinary.com/pensionbee/video/upload/q_auto/v1646048083/Pensions%20Academy/pension-academy-videos/Pension_Academy_Ep_1_CC.mp4"
        thumbnail="https://res.cloudinary.com/pensionbee/image/upload/v1644591246/Pensions%20Academy/thumbnail_1.jpg"
      />
    </GridContainer>
  );
};

export const Modal = () => {
  return (
    <GridContainer>
      <Video
        src="https://res.cloudinary.com/pensionbee/video/upload/q_auto/v1646048083/Pensions%20Academy/pension-academy-videos/Pension_Academy_Ep_1_CC.mp4"
        thumbnail="https://res.cloudinary.com/pensionbee/image/upload/v1644591246/Pensions%20Academy/thumbnail_1.jpg"
        isModal
      />
    </GridContainer>
  );
};

export const InColumn = () => {
  return (
    <GridContainer>
      <GridRow>
        <GridItem span={{ lg: 4 }}>
          <Box backgroundColor="grey4" h="100%" />
        </GridItem>
        <GridItem span={{ lg: 4 }}>
          <Video
            src="https://res.cloudinary.com/pensionbee/video/upload/q_auto/v1646048083/Pensions%20Academy/pension-academy-videos/Pension_Academy_Ep_1_CC.mp4"
            thumbnail="https://res.cloudinary.com/pensionbee/image/upload/v1644591246/Pensions%20Academy/thumbnail_1.jpg"
          />
        </GridItem>
        <GridItem span={{ lg: 4 }}>
          <Box backgroundColor="grey4" h="100%" />
        </GridItem>
      </GridRow>
    </GridContainer>
  );
};

export const Muted = () => {
  return (
    <GridContainer>
      <Video
        src="https://res.cloudinary.com/pensionbee/video/upload/q_auto/v1646048083/Pensions%20Academy/pension-academy-videos/Pension_Academy_Ep_1_CC.mp4"
        thumbnail="https://res.cloudinary.com/pensionbee/image/upload/v1644591246/Pensions%20Academy/thumbnail_1.jpg"
        muted={true}
      />
    </GridContainer>
  );
};

export const NoThumbnail = () => {
  return (
    <GridContainer>
      <Video src="https://res.cloudinary.com/pensionbee/video/upload/q_auto/v1646048083/Pensions%20Academy/pension-academy-videos/Pension_Academy_Ep_1_CC.mp4" />
    </GridContainer>
  );
};
