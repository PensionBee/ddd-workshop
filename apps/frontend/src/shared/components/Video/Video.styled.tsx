import { Box } from "@shared/components/Box";
import styled from "styled-components";

export const StyledIconWrapper = styled(Box)`
  transform: translate(-50%, -50%);
`;

export const StyledPlayButton = styled.button`
  cursor: pointer;
`;
