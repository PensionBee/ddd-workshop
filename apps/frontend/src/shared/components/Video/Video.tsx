import { LargeSystemIcon } from "../Svg/Svg";
import { Box } from "../Box/Box";
import { useState } from "react";
import type { ReactPlayerProps } from "react-player/lazy";
import ReactPlayer from "react-player/lazy";
import { Modal } from "../Modal/Modal";
import { StyledPlayButton, StyledIconWrapper } from "./Video.styled";

export type VideoProps = {
  /** The url path to the video */
  src: string;
  /** Overrides the default video thumbnail */
  thumbnail?: string;
  /** Mutes the video player */
  muted?: ReactPlayerProps["muted"];
  /** The video controls */
  controls?: ReactPlayerProps["controls"];
  /** Set to true or false to loop the media */
  loop?: ReactPlayerProps["loop"];
  /** If in lazy mode */
  fallback?: ReactPlayerProps["fallback"];
  /** If the video should play in a modal */
  isModal?: boolean;
};

export const Video: React.FC<VideoProps> = ({
  src,
  thumbnail = "",
  muted = false,
  controls = true,
  loop = false,
  fallback,
  isModal = false,
  ...props
}) => {
  const [isVideoModalOpen, setIsVideoModalOpen] = useState(false);
  const [isInlineVideoOpen, setIsInlineVideoOpen] = useState(false);

  if (isModal) {
    return (
      <Box position="relative" w="100%" pb="56.25%">
        <Box position="absolute" top={0} left={0} right={0} bottom={0}>
          <StyledPlayButton onClick={() => setIsVideoModalOpen(true)}>
            <StyledIconWrapper position="absolute" top="50%" left="50%">
              <LargeSystemIcon icon="Play" w={64} h={64} />
            </StyledIconWrapper>
            <img src={thumbnail} />
          </StyledPlayButton>
        </Box>
        <Modal
          isOpen={isVideoModalOpen}
          handleClose={() => setIsVideoModalOpen(false)}
          backgroundColor="transparent"
          wMax="100vw"
          hMax="100vh"
        >
          <Box d="flex">
            <ReactPlayer
              url={src}
              muted={muted}
              controls={controls}
              loop={loop}
              fallback={fallback}
              playing
              {...props}
            />
          </Box>
        </Modal>
      </Box>
    );
  }

  return (
    <Box position="relative" w="100%" pb="56.25%">
      <Box position="absolute" top={0} left={0} right={0} bottom={0}>
        {!isInlineVideoOpen ? (
          <StyledPlayButton onClick={() => setIsInlineVideoOpen(true)}>
            <StyledIconWrapper position="absolute" top="50%" left="50%">
              <LargeSystemIcon icon="Play" w={64} h={64} />
            </StyledIconWrapper>
            <img src={thumbnail} />
          </StyledPlayButton>
        ) : (
          <ReactPlayer
            url={src}
            muted={muted}
            controls={controls}
            loop={loop}
            fallback={fallback}
            width="100%"
            height="100%"
            playing
            {...props}
          />
        )}
      </Box>
    </Box>
  );
};
