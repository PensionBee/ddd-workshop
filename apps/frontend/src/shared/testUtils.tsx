import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
