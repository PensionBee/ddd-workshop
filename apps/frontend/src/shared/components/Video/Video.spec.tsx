import { fireEvent, render } from "@shared/testUtils";
import { act } from "react-dom/test-utils";
import type { ReactPlayerProps } from "react-player";
import { Video } from "./Video";

const COMPONENT = "video-component";
const EXAMPLE_URL = "example-url";
const EXAMPLE_THUMB = "example-thumbnail";

const CLICK_EVENT = new MouseEvent("click", {
  bubbles: true,
  cancelable: true,
});

jest.mock("react-player/lazy", () => ({
  __esModule: true,
  default: ({ url, muted, controls, loop, fallback }: ReactPlayerProps) => (
    <div
      data-url={url}
      data-muted={muted}
      data-controls={controls}
      data-loop={loop}
      data-fallback={fallback}
    >
      {COMPONENT}
    </div>
  ),
}));

describe("Video", () => {
  describe("Default", () => {
    it("should render correctly", () => {
      const { asFragment, queryByText, container } = render(
        <Video src={EXAMPLE_URL} thumbnail={EXAMPLE_THUMB} />
      );
      const button = container.querySelector("button");
      expect(queryByText(COMPONENT)).not.toBeInTheDocument();
      act(() => button && fireEvent(button, CLICK_EVENT));
      expect(queryByText(COMPONENT)).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render correctly without thumbnail", () => {
      const { asFragment, queryByText, container } = render(
        <Video src={EXAMPLE_URL} />
      );
      const button = container.querySelector("button");
      expect(queryByText(COMPONENT)).not.toBeInTheDocument();
      act(() => button && fireEvent(button, CLICK_EVENT));
      expect(queryByText(COMPONENT)).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("Modal", () => {
    it("should render correctly", () => {
      const { asFragment, queryByText, container } = render(
        <Video src={EXAMPLE_URL} thumbnail={EXAMPLE_THUMB} isModal />
      );
      const button = container.querySelector("button");
      expect(queryByText(COMPONENT)).not.toBeInTheDocument();
      act(() => button && fireEvent(button, CLICK_EVENT));
      expect(queryByText(COMPONENT)).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should close correctly", () => {
      const { asFragment, queryByText, container } = render(
        <Video src={EXAMPLE_URL} thumbnail={EXAMPLE_THUMB} isModal />
      );
      const button = container.querySelector("button");
      expect(button).toBeInTheDocument();

      act(() => button && fireEvent(button, CLICK_EVENT));
      expect(queryByText(COMPONENT)).toBeInTheDocument();

      const closeButton = document.querySelector("#modal-root button");
      expect(closeButton).toBeInTheDocument();

      act(() => closeButton && fireEvent(closeButton, CLICK_EVENT));
      expect(queryByText(COMPONENT)).not.toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
