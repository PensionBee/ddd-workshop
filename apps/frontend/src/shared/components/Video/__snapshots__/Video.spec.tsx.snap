// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Video > Default > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm jUKRYu"
  >
    <div
      class="sc-gsnScm ezwATA"
    >
      <div
        data-controls="true"
        data-loop="false"
        data-muted="false"
        data-url="example-url"
      >
        video-component
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video > Default > should render correctly without thumbnail 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm jUKRYu"
  >
    <div
      class="sc-gsnScm ezwATA"
    >
      <div
        data-controls="true"
        data-loop="false"
        data-muted="false"
        data-url="example-url"
      >
        video-component
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video > Modal > should close correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm jUKRYu"
  >
    <div
      class="sc-gsnScm ezwATA"
    >
      <button
        class="sc-jSMdHm bFHbXQ"
      >
        <div
          class="sc-gsnScm fKVHzT sc-eCYdKt KsdpT"
        >
          <span
            class="sc-bczSft gbfcOk"
            data-rc="LargeSystemIcon.Play"
          >
            <svg
              fill="none"
              height="32"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M23.2565 17.5126C24.2463 16.6758 24.2493 15.3216 23.2565 14.4823L9.7923 3.10005C8.80245 2.26326 8.00001 2.67968 8.00001 3.99293V28.002C8.00001 29.3319 8.79944 29.7342 9.7923 28.8948L23.2565 17.5126Z"
                fill="#F9C000"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
        <img
          src="example-thumbnail"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video > Modal > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm jUKRYu"
  >
    <div
      class="sc-gsnScm ezwATA"
    >
      <button
        class="sc-jSMdHm bFHbXQ"
      >
        <div
          class="sc-gsnScm fKVHzT sc-eCYdKt KsdpT"
        >
          <span
            class="sc-bczSft gbfcOk"
            data-rc="LargeSystemIcon.Play"
          >
            <svg
              fill="none"
              height="32"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M23.2565 17.5126C24.2463 16.6758 24.2493 15.3216 23.2565 14.4823L9.7923 3.10005C8.80245 2.26326 8.00001 2.67968 8.00001 3.99293V28.002C8.00001 29.3319 8.79944 29.7342 9.7923 28.8948L23.2565 17.5126Z"
                fill="#F9C000"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
        <img
          src="example-thumbnail"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR jicaaU"
  >
    <div
      class="sc-gswNZR gJxzFG"
    >
      <div
        data-controls="true"
        data-loop="false"
        data-muted="false"
        data-url="example-url"
      >
        video-component
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video Default should render correctly without thumbnail 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR jicaaU"
  >
    <div
      class="sc-gswNZR gJxzFG"
    >
      <div
        data-controls="true"
        data-loop="false"
        data-muted="false"
        data-url="example-url"
      >
        video-component
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video Modal should close correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR jicaaU"
  >
    <div
      class="sc-gswNZR gJxzFG"
    >
      <button
        class="sc-jSUZER iXDapD"
      >
        <div
          class="sc-gswNZR ijzHa-d sc-eDvSVe hSqznv"
        >
          <span
            class="sc-bcXHqe ffJoEj"
            data-rc="LargeSystemIcon.Play"
          >
            <svg>
              play.svg
            </svg>
          </span>
        </div>
        <img
          src="example-thumbnail"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video Modal should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR jicaaU"
  >
    <div
      class="sc-gswNZR gJxzFG"
    >
      <button
        class="sc-jSUZER iXDapD"
      >
        <div
          class="sc-gswNZR ijzHa-d sc-eDvSVe hSqznv"
        >
          <span
            class="sc-bcXHqe ffJoEj"
            data-rc="LargeSystemIcon.Play"
          >
            <svg>
              play.svg
            </svg>
          </span>
        </div>
        <img
          src="example-thumbnail"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
